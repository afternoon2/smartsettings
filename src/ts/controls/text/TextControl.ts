import { Control } from '../control/Control';
import { InternalState, TextControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/text.sass';

export type TextControlProps = {
  id: string,
  options: TextControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'text';

  protected static template = (state: InternalState): string => `
    <input
      type="text"
      class="${Styles.text}"
      ${state.disabled ? 'disabled' : ''}
      ${state.value ? `value="${state.value}"` : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `min-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.pattern ? `pattern="${state.pattern}"` : ''}
      ${state.readOnly ? 'readonly="true"' : ''}
    />
  `;

  constructor(props: TextControlProps) {
    super({
      ...props,
      template: TextControl.template
    });
    this.controlElement = this.element.querySelector('[type="text"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
