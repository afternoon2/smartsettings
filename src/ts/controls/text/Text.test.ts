import { TextControl } from './TextControl';

let control: TextControl;

const listener = jest.fn();

describe('Text control', () => {
  beforeEach(() => {
    control = new TextControl({
      id: 'id',
      options: {
        name: 'name',
        value: 'test',
        autocomplete: false,
        disabled: false,
        listener,
      },
      parentElement: document.body,
    });
  });

  test('If it matches snapshot after initial load', () => {
    expect(control.element).toMatchSnapshot();
  });

  test('Disable', () => {
    control.disable();
    expect(control.element).toMatchSnapshot();
  });

  test('Enable', () => {
    control.disable();
    control.enable();
    expect(control.controlElement.hasAttribute('disabled')).toBe(false);
    expect(control.element).toMatchSnapshot();
  });

  test('Get and set value', () => {
    control.value = 'Test';
    expect(control.value).toBe('Test');
  });

  test('If it assings user listener correctly', () => {
    expect(listener).toHaveBeenCalledTimes(4);
  });
});