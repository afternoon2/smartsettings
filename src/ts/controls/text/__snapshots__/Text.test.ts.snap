// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text control Disable 1`] = `
<div
  class="control disabled"
  disabled="true"
>
  
      
  <main
    class="control__body"
    id="id"
  >
    
        
    <p
      class="control__name"
      title="name"
    >
      
          name
        
    </p>
    
      
  </main>
  
      
  <div
    class="control__content"
  >
    
        
    
    <input
      autofocus="true"
      class="text"
      type="text"
      value="test"
    />
    
  
      
  </div>
  
  
</div>
`;

exports[`Text control Enable 1`] = `
<div
  class="control"
  disabled="true"
>
  
      
  <main
    class="control__body"
    id="id"
  >
    
        
    <p
      class="control__name"
      title="name"
    >
      
          name
        
    </p>
    
      
  </main>
  
      
  <div
    class="control__content"
  >
    
        
    
    <input
      autofocus="true"
      class="text"
      type="text"
      value="test"
    />
    
  
      
  </div>
  
  
</div>
`;

exports[`Text control If it matches snapshot after initial load 1`] = `
<div
  class="control"
>
  
      
  <main
    class="control__body"
    id="id"
  >
    
        
    <p
      class="control__name"
      title="name"
    >
      
          name
        
    </p>
    
      
  </main>
  
      
  <div
    class="control__content"
  >
    
        
    
    <input
      autofocus="true"
      class="text"
      type="text"
      value="test"
    />
    
  
      
  </div>
  
  
</div>
`;
